src/signTxPoolRegistration.c

```diff
@@ -406,16 +895,182 @@ static void signTxPoolRegistration_handleOwnerAPDU(uint8_t* wireDataBuffer, size
 // ============================== RELAY ==============================
 
 enum {
-	RELAY_SINGLE_HOST_IP = 0,
-	RELAY_SINGLE_HOST_NAME = 1,
-	RELAY_MULTIPLE_HOST_NAME = 2
+	HANDLE_RELAY_IP_STEP_DISPLAY_NUMBER = 6700,
+	HANDLE_RELAY_IP_STEP_DISPLAY_IPV4,
+	HANDLE_RELAY_IP_STEP_DISPLAY_IPV6,
+	HANDLE_RELAY_IP_STEP_DISPLAY_PORT,
+	HANDLE_RELAY_IP_STEP_RESPOND,
+	HANDLE_RELAY_IP_STEP_INVALID,
 };
 
+static void handleRelay_ip_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = handleRelay_ip_ui_runStep;
+
+	pool_relay_t* relay = &subctx->stateData.relay;
+
+	UI_STEP_BEGIN(subctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_RELAY_IP_STEP_DISPLAY_NUMBER) {
+		ui_displayRelaycreen(
+			relay,
+			subctx->currentRelay,
+			this_fn
+		);
+	}
+	UI_STEP(HANDLE_RELAY_IP_STEP_DISPLAY_IPV4) {
+		ui_displayIpv4Screen(
+		        &relay->ipv4,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_RELAY_IP_STEP_DISPLAY_IPV6) {
+		ui_displayIpv6Screen(
+		        &relay->ipv6,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_RELAY_IP_STEP_DISPLAY_PORT) {
+		ui_displayIpPortScreen(
+		        &relay->port,
+		        this_fn
+		);
+	}
+	UI_STEP(HANDLE_RELAY_IP_STEP_RESPOND) {
+		respondSuccessEmptyMsg();
+
+		subctx->currentRelay++;
+		TRACE("current relay %d", subctx->currentRelay);
+
+		if (subctx->currentRelay == subctx->numRelays) {
+			advanceState();
+		}
+	}
+	UI_STEP_END(HANDLE_RELAY_IP_STEP_INVALID);
+}
+
 enum {
-	RELAY_NO = 1,
-	RELAY_YES = 2
+	HANDLE_RELAY_DNS_STEP_DISPLAY_NUMBER = 6800,
+	HANDLE_RELAY_DNS_STEP_DISPLAY_PORT,
+	HANDLE_RELAY_DNS_STEP_DISPLAY_DNSNAME,
+	HANDLE_RELAY_DNS_STEP_RESPOND,
+	HANDLE_RELAY_DNS_STEP_INVALID,
 };
 
+static void handleRelay_dns_ui_runStep()
+{
+	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
+	ui_callback_fn_t* this_fn = handleRelay_dns_ui_runStep;
+
+	pool_relay_t* relay = &subctx->stateData.relay;
+
+	UI_STEP_BEGIN(subctx->ui_step, this_fn);
+
+	UI_STEP(HANDLE_RELAY_DNS_STEP_DISPLAY_NUMBER) {
+		ui_displayRelaycreen(
+			relay,
+			subctx->currentRelay,
+			this_fn
+		);
+	}
+	UI_STEP(HANDLE_RELAY_DNS_STEP_DISPLAY_PORT) {
```

Comment:
It feels a bit weird to show the port first and then the DNS record name - I'd invert it